xdsl/interpreters/riscv.py

```diff
@@ -0,0 +1,217 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Any, Callable, Generic, Iterator, TypeAlias, TypeVar
+
+from xdsl.dialects import riscv
+from xdsl.dialects.builtin import AnyIntegerAttr, IntegerAttr, ModuleOp
+from xdsl.interpreter import (
+    Interpreter,
+    InterpreterFunctions,
+    PythonValues,
+    impl,
+    register_impls,
+)
+from xdsl.interpreters.comparisons import to_signed, to_unsigned
+from xdsl.ir.core import Operation
+from xdsl.utils.exceptions import InterpretationError
+
+_T = TypeVar("_T")
+
+
+def pairs(els: list[_T]) -> Iterator[tuple[_T, _T]]:
+    count = len(els)
+    assert not count % 2
+    for i in range(0, count, 2):
+        yield els[i], els[i + 1]
+
+
+CustomInstructionFn: TypeAlias = Callable[
+    [Interpreter, riscv.CustomAssemblyInstructionOp, PythonValues], PythonValues
+]
+
+
+@dataclass
+class Buffer(Generic[_T]):
+    """
+    Data structure to help simulate pointer offsets into buffers of data in memory.
+    """
+
+    data: list[_T]
+    offset: int = 0
+
+    def __add__(self, offset: int) -> Buffer[_T]:
+        """
+        Aliases the data list, so storing into the offset stores for all other references
+        to the list.
+        """
+        return Buffer(self.data, self.offset + offset // 4)
```

Comment:
I don't mind too much but feels strange to have a generic data structure and a hardcoded "size" constant here.